src/DeterministicNoiseSimulator.cpp

```diff
@@ -4,680 +4,77 @@
-                    //todo I only must check array elements <=current_v, for the caching
-                    sort(used_qubits.begin(), used_qubits.end(), std::greater<>());
-
-                    [[maybe_unused]] auto cache_size_before = dd->cn.cacheCount();
-
-                    dd::Edge<dd::dNode> nodeAfterNoise = {};
-                    if (use_density_matrix_type) {
-                        dEdge::applyDmChangesToEdges(&density_root_edge, nullptr);
-                        nodeAfterNoise = applyNoiseEffects(density_root_edge, used_qubits, false);
-                        dEdge::revertDmChangesToEdges(&density_root_edge, nullptr);
-                    } else {
-                        nodeAfterNoise = applyNoiseEffects(density_root_edge, used_qubits, true);
-                    }
-
-                    [[maybe_unused]] auto cache_size_after = dd->cn.cacheCount();
-                    assert(cache_size_after == cache_size_before);
-
-                    dd->incRef(nodeAfterNoise);
-
-                    dEdge::alignDensityEdge(&density_root_edge);
-                    dd->decRef(density_root_edge);
-                    density_root_edge = nodeAfterNoise;
-                    if (use_density_matrix_type) {
-                        dEdge::setDensityMatrixTrue(&density_root_edge);
-                    }
-                }
-            }
+            deterministicNoiseFunctionality.applyNoiseEffects(rootEdge, op);
         }
     }
-
-    dd->garbageCollect(true);
-    printf("Multiplication cache");
-    dd->densityDensityMultiplication.printStatistics();
-    printf("Add cache\n");
-    dd->densityAdd.printStatistics();
-    printf("Noise cache\n");
-    dd->densityNoise.printStatistics();
-    return AnalyseState(n_qubits, false);
+    return Simulator<DDPackage>::dd->getProbVectorFromDensityMatrix(rootEdge, measurementThreshold);
 }
 
-dEdge DeterministicNoiseSimulator::applyNoiseEffects(dEdge& originalEdge, const std::vector<dd::Qubit>& used_qubits, bool firstPathEdge) {
-    if (originalEdge.p->v < used_qubits.back() || originalEdge.isTerminal()) {
-        dEdge tmp{};
-        if (originalEdge.w.approximatelyZero()) {
-            tmp.w = dd::Complex::zero;
-        } else {
-            tmp.w = originalEdge.w;
-            //            noiseLookUpResult.w = dd->cn.getCached(dd::CTEntry::val(originalEdge.w.r), dd::CTEntry::val(originalEdge.w.i));
-        }
-        if (originalEdge.isTerminal()) {
-            return dEdge::terminal(tmp.w);
-        } else {
-            tmp.p = originalEdge.p;
-            return tmp;
-        }
+template<class DDPackage>
+std::map<std::string, std::size_t> DeterministicNoiseSimulator<DDPackage>::sampleFromProbabilityMap(const std::map<std::string, dd::fp>& resultProbabilityMap, unsigned int shots) {
+    // Create probability distribution from measure probabilities
+    std::vector<dd::fp> weights;
+    for (auto& res: resultProbabilityMap) {
+        weights.push_back(res.second);
     }
```

Comment:
use `const` and structured bindings to make the intent clearer
```suggestion
    for (const auto& [state, prob]: resultProbabilityMap) {
        weights.push_back(prob);
    }
```